### DISCUSSION TITLE
Blazor 6.0 Call BusinessRule On business object Delete Method

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#3240] [Questions] [answered]
Hello All,

Sorry all Im not sure how to format it 

I need some help with this. I will provide some info on what I am trying to do. The application that I am working on tracks equipment. When the admin tries to delete the equipment, I need to check if it has been assigned to anyone. If the equipment is assigned, then the rule should break and not allow the delete. 
What I notice is that when the Delete method is call the IsDeleted property is false; So I call the base.MarkDeleted() to set the flag. I then call the business rule by property so not all are called on delete. The issue is that the rule does not execute the constructor of my business rule class and I am not able to pass base.IsDelete to help me apply the rule. Is there a way to pass this value on Delete? Below is my code sample. 

**Business Base** 

```
        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new CheckIfAssignedBeforDelete(EquipmentIDProperty, base.IsDeleted));
        }
```

```
        [Delete]
        private void Delete(Guid EquipmentID, [Inject] DataAccess.IEquipmentDal dal)
        {
            base.MarkDeleted();
            BusinessRules.CheckRules(EquipmentIDProperty);
            dal.Delete(EquipmentID);
        }

```

**Business Rule**

```
    public class CheckIfAssignedBeforDelete : BusinessRule
    {
        bool IsMarkForDelete = false;

        public CheckIfAssignedBeforDelete(Csla.Core.IPropertyInfo EquipmentIDProperty,
                                bool isMarkDeleted)
          : base(EquipmentIDProperty)
        {
            //Note Order make a difference with InputProperties
            PrimaryProperty = EquipmentIDProperty;
            IsMarkForDelete = isMarkDeleted;
        }

        protected override void Execute(IRuleContext context)
        {
            if (this.IsMarkForDelete)
            {
                Guid equipID = (Guid)ReadProperty(context.Target, PrimaryProperty);
                var obj1 = new EmployeeEquipmentExistsCommand();
                obj1.ID = equipID;

                if(obj1 != null && obj1.EmployeeEquipmentExists)
                {
                    context.AddErrorResult("The equipment cannot be deleted since it has been assigned to an employee.");
                }
            }
        }
    }


```

**Comment by jhuerta3** [comment]
Well I added this code and it seemed to work. Im not sure if this is the best method to do it though. Since I know the business rules work with properties, I added a MarkForDelete property to my business object. That was the only way I figured out how to pass the boolean value. 

Can someone let me know this is a proper way of accomplishing my task.


```
        public static readonly PropertyInfo<bool> MarkForDeleteProperty = RegisterProperty<bool>(nameof(MarkForDelete));
        [Display(Name = "MarkForDelete")]
        public bool MarkForDelete
        {
            get { return GetProperty(MarkForDeleteProperty); }
            set { SetProperty(MarkForDeleteProperty, value); }
        }
```


```
        [Delete]
        private void Delete(Guid equipmentID, [Inject] DataAccess.IEquipmentDal dal)
        {
            base.MarkDeleted();
            EquipmentID = equipmentID;
            MarkForDelete = base.IsDeleted;
            BusinessRules.CheckRules(EquipmentIDProperty);
            if (IsValid)
            {
                dal.Delete(EquipmentID);
            }
        }
```

**Reply by rockfordlhotka** [reply]
This is probably not what you want, because the `Delete` operation method is running on the server and so can't proactively tell the user that they can't delete the object.

Or maybe it is what you want - just keep in mind that if a delete isn't possible then you should throw an exception so the failure flows back to the client so it can be reported to the user.

**Comment by rockfordlhotka** [comment]
There are two ways to implement a delete operation on a root object.

One is to use the data portal `DeleteAsync` method, which usually passes an id value to your server-side data portal `Delete` operation method, and that method deletes the data corresponding to the id.

The other is probably what you want, and works like this:

1. Fetch the object
2. Usually show the data to the user, let them decide to delete or not
3. Run any rules to see if the object can be deleted
4. Call `MarkDelete` in the object (this is a protected method) to mark the object for deletion (`IsDeleted ==true`)
5. Call `SaveAsync` on the object to "save" it - which means it'll ultimately call the `DeleteSelf` server-side operation because `IsDeleted` is true

**Comment by Chicagoan2016** [comment]
Could we possibly add a gateway rule that will only check for equipment assignment when the object is marked for deletion?
I was wondering if calling obj.Delete() is equivalent to step 4 in @rockfordlhotka reply?
Kind Regards

**Comment by rockfordlhotka** [comment, marked as answer]
I think this is probably the right answer?

```c#
using Csla;
using Csla.Configuration;
using Csla.Core;
using Csla.Rules;
using Microsoft.Extensions.DependencyInjection;

// configuration

var services = new ServiceCollection();
services.AddCsla();
var provider = services.BuildServiceProvider();

// test code

var dp = provider.GetRequiredService<IDataPortal<EquipmentDelete>>();
var equipment = dp.Fetch(42);
try
{
  var deleted = await equipment.RemoveEquipment();
  if (deleted)
  {
    Console.WriteLine("Deleted successfully");
  }
  else
  {
    Console.WriteLine(equipment.BrokenRulesCollection.GetFirstBrokenRule(EquipmentDelete.IdProperty).Description);
  }
}
catch (DataPortalException ex)
{
  if (ex.BusinessException != null)
    Console.WriteLine($"Error: {ex.BusinessException.ToString()}");
  else
    Console.WriteLine($"Unexpected error: {ex.ToString()}");
}
catch (Exception ex)
{
  Console.WriteLine($"Unexpected error: {ex.ToString()}");
}

// classes follow here

[Serializable]
public class EquipmentDelete : BusinessBase<EquipmentDelete>
{
  public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(nameof(Id));
  public int Id
  {
    get => GetProperty(IdProperty);
    set => SetProperty(IdProperty, value);
  }

  public async Task<bool> RemoveEquipment()
  {
    await BusinessRules.CheckRulesAsync();
    if (IsValid)
    {
      MarkDeleted();
      await SaveAsync();
      return true;
    }
    else
    {
      return false;
    }
  }

  protected override void AddBusinessRules()
  {
    base.AddBusinessRules();
    BusinessRules.AddRule(new EquipmentAssignedRuleAsync(IdProperty));
  }

  [Fetch]
  private void Fetch(int id)
  {
    // get data related to `id`
    using (BypassPropertyChecks)
    {
      Id = id;
    }
  }

  [Insert]
  private void Insert()
  {
    // insert new data related to `id`
  }

  [Update]
  private void Update() 
  { 
    // update data related to `id`
  }

  [DeleteSelf]
  private void DeleteSelf()
  {
    Delete(Id);
  }

  [Delete]
  private void Delete(int id)
  {
    // delete data related to `id`
  }
}

public class EquipmentAssignedRuleAsync : BusinessRuleAsync
{
  public EquipmentAssignedRuleAsync(IPropertyInfo primaryProperty)
    : base(primaryProperty) 
  { 
    InputProperties.Add(PrimaryProperty);
  }

  protected override async Task ExecuteAsync(IRuleContext context)
  {
    var id = (int)context.InputPropertyValues[PrimaryProperty];
    var dp = context.ApplicationContext.GetRequiredService<IDataPortal<EquipmentAssignedChecker>>();
    var checker = await dp.FetchAsync(id);
    if (checker.IsAssigned)
      context.AddErrorResult("Equipment is assigned to someone");
  }
}

[Serializable]
public class EquipmentAssignedChecker : ReadOnlyBase<EquipmentAssignedChecker>
{
  public static readonly PropertyInfo<bool> IsAssignedProperty = RegisterProperty<bool>(nameof(IsAssigned));
  public bool IsAssigned
  {
    get => GetProperty(IsAssignedProperty);
    set => LoadProperty(IsAssignedProperty, value);
  }
  [Fetch]
  private void Fetch(int id)
  {
    // find out if `id` equipment is assigned
    IsAssigned = false;
  }
}
```

**Reply by rockfordlhotka** [reply]
Put this into `Program.cs` of a console app. You can change the `IsAssigned = false` to `IsAssigned = true` in the last line of code to see the object not save b/c it is assigned.

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka and @Chicagoan2016,

Thank you for the responses and code snipes. I totally forgot about the second method of deleting. I will also look at implementing the option above. I have been struggling to get the broken rules back to the client with my original code that I added for the delete. I will try it out and get back to you. 

Thank you both.

**Comment by michaelcsikos** [comment]
As the object is fetched you could load a value `HasBeenAssigned` which would allow you to disable the delete option in your UI. Loading the `NumberOfAssignments` could be better still as you could display this. You could also still run the exists command when `MarkForDelete` is set to `true` to check if the equipment has been assigned in the time since the object was loaded (a slim chance), and the final fallback is throwing an exception in the Data Portal, as Rocky said. This exception could happen automatically if you have foreign key constraints which prevent your object from being deleted. The SQL error message can be a bit cryptic for users, though, e.g. `violation of foreign key constraint FK_...`.

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka and @Chicagoan2016,

Well Im stilling trying to perform the delete and check my rules. I am trying to perform it via this method.

Fetch the object
Usually show the data to the user, let them decide to delete or not
Run any rules to see if the object can be deleted
Call MarkDelete in the object (this is a protected method) to mark the object for deletion (IsDeleted ==true)
Call SaveAsync on the object to "save" it - which means it'll ultimately call the DeleteSelf server-side operation because IsDeleted is true

I know you guys provided the other solution but I want to see if I can get it going. 


**Razor Page Code**

```
<button disabled="@(vm.Model.IsNew)" class="btn btn-secondary" @onclick="() => SelectForDelete(vm.Model.EquipmentID.ToString())">Delete</button>

    private void SelectForDelete(string id)
    {
        itemSelectedForDeletion = id;
        vm.Model.MarkForDelete = true;
    }
```



Im getting the following error below with the following code on my command. 

`rule://assetmgmtsystem.shared.businessrules.checkifassignedbefordelete/EquipmentID:One or more errors occurred. (Serialization of type object is not supported. (System.Object[]))`

![image](https://user-images.githubusercontent.com/56940135/211661322-4090840e-fc33-40ad-9140-f1aeb819b320.png)

**Command Execute**
```
        protected override void Execute(IRuleContext context)
        {
            Guid equipID = (Guid)ReadProperty(context.Target, PrimaryProperty);
            IsMarkForDelete = (bool)context.InputPropertyValues[InputProperties[1]];

            if (Guid.Empty != equipID && IsMarkForDelete)
            {

                //This code below is causing my error
                var dataPortal = context.DataPortalFactory.GetPortal<EmployeeEquipmentExistsCommand>();
                var obj1 = dataPortal.CreateAsync(equipID);
                var obj2 = dataPortal.ExecuteAsync(obj1);
                var bolv = obj2.Result.EmployeeEquipmentExists;
                // *****

                if (equipID != Guid.Empty && bolv)
                {
                context.AddErrorResult("The equipment cannot be deleted since it has been assigned to an employee.");
                }
            }
        }
```

If I comment out the dataPortal create and execute code and if statement I do see my error message. Is there something that I can do to correct the error?

If not then I guess I will need to implement the solution above. 


I wish I could just call my DataAccess layer class directly 

`public class EmployeeEquipmentEFDal : IEmployeeEquipmentDa`

```
        public bool Exists(Guid equipmentID)
        {
            var data = _context.EmployeeEquipments.Where(
                x => x.EquipmentID == equipmentID).FirstOrDefault();
            return !(data == null);
        }

```

**Comment by Chicagoan2016** [comment]
@jhuerta3 
I am not sure about the code you mentioned is causing error. Could you run it successfully outside the rule?

**Reply by jhuerta3** [reply]
@Chicagoan2016 I think you are correct. I think there might be something else going on. I will do more research and confirm. 
Thank you. If not Im going with the sample that Rocky gave me.

**Comment by jhuerta3** [comment]
Thank you @rockfordlhotka and @Chicagoan2016.
I impletemented the EquipmentDelete sample that you provided and that worked. I created a ErrorMessage to show the error on the UI. 

Thank you. 


**On Razor Page**
```

<ErrorMessage Error="@ErrorMessage" ErrorType="@ErrorMessageType" />

<button @onclick="() => Delete(vm.Model.EquipmentID.ToString())"> <span class="alert-danger">Confirm delete</span></button>


    private async void Delete(string id)
    {
        itemSelectedForDeletion = "0";
        ErrorMessage = "";
        ErrorMessageType = "";

        var equipment = await EquipmentDeletePortal.FetchAsync(new Guid(id));
        try
        {
            var deleted = await equipment.RemoveEquipment();
            if (deleted)
            {
                NavManager.NavigateTo("/EquipmentList");
            }
            else
            {
                vm.Model.EquipmentID = equipment.Id;
                ErrorMessage = equipment.BrokenRulesCollection.GetFirstBrokenRule(EquipmentDelete.IdProperty).Description;
                ErrorMessageType = "danger";
                StateHasChanged();
            }
        }
        catch (Csla.DataPortalException ex)
        {
            if (ex.BusinessException != null){
                ErrorMessage = $"Error: {ex.BusinessException.ToString()}";
                ErrorMessageType = "danger";
                StateHasChanged();
            }
            else
            {
                ErrorMessage = $"Unexpected error: {ex.ToString()}";
                ErrorMessageType = "danger";
                StateHasChanged();
            }
        }
        catch (Exception ex)
        {
            ErrorMessage = $"Unexpected error: {ex.ToString()}";
            ErrorMessageType = "danger";
            StateHasChanged();
        }
    }

```
